shared_stack.bash

```diff
@@ -0,0 +1,214 @@
+# Top-level product to install
+PRODUCT=lsst_distrib
+# Number of daily releases to keep
+# shellcheck disable=SC2034
+KEEP_d=24
+# Number of weekly releases to keep
+# shellcheck disable=SC2034
+KEEP_w=26
+
+#######
+# Setup
+#######
+
+while getopts n opt; do
+  case "$opt" in
+    n) dryrun="echo"
+       ;;
+    *) echo "Usage: $0 [-n]"
+       exit 1 ;;
+  esac
+done
+[ $OPTIND -le $# ] && echo "Usage: $0 [-n]" && exit 1
+# Ensure that there's no contamination from a pre-existing environment
+unset LSST_CONDA_ENV_NAME PYTHONPATH
+
+##############
+# Installation
+##############
+
+set -e
+mkdir -p "$STATE"
+cd "$STATE"
+# Label temporaries with our unique PID
+tmp=tmp.$$
+mkdir -p "$tmp" "$ROOT/tag"
+[ -e tags.deleted ] || touch tags.deleted
+umask 022
+
+# Get list of tags
+$dryrun curl -L https://ls.st/lsstinstall -o $tmp/lsstinstall
+$dryrun mv $tmp/lsstinstall lsstinstall
+curl -L https://eups.lsst.codes/stack/src/tags | while read -r line; do
+  [[ "$line" =~ \>([dvw][0-9_]+(rc[0-9]+)?)\.list\< ]] && echo "${BASH_REMATCH[1]}"
+done | sort > $tmp/tags
+
+for tag in $(comm -1 -3 <(sort tags.deleted) $tmp/tags); do
+  dir="$ROOT/tag/$tag"
+  [ -d "$dir" ] && continue
+  $dryrun mkdir -p "${dir}.$tmp" || continue
+  (
+     set -e
+     $dryrun cd "${dir}.$tmp"
+     $dryrun bash "$STATE/lsstinstall" -T "$tag" -p "$ROOT"/conda -S
+     set +x
+     $dryrun source loadLSST.sh
+     set -x
+     # Install the developer add-on packages if needed
+     # When rubin-env-developer becomes available, replace with this:
+     # conda list --json | grep rubin-env-developer > /dev/null \
+     #   || mamba install rubin-env-developer
+     if [ -n "$dryrun" ]; then
+       echo "conda list --json | grep mypy > /dev/null || mamba install ..."
+     else
+       # shellcheck disable=SC2046
+       conda list --json | grep mypy > /dev/null \
+         || mamba install -y -c conda-forge --no-update-deps $(cat "$STATE/developer.txt")
+     fi
+     $dryrun eups distrib install -t "$tag" "$PRODUCT"
+
+     # Rewrite load scripts and write useful files
+     $dryrun cp loadLSST.sh tmp.loadLSST.sh
+     if [ -n "$dryrun" ]; then
+       echo "echo . $ROOT/loadLSST.sh ... > loadLSST.sh"
+       echo "echo conda activate ... > switch.sh"
+       echo "echo ... > env_name"
+     else
+       echo "LSST_CONDA_ENV_NAME=$LSST_CONDA_ENV_NAME . \"$ROOT/loadLSST.sh\"" > loadLSST.sh
+       echo "conda activate $LSST_CONDA_ENV_NAME" > switch.sh
+       echo "$LSST_CONDA_ENV_NAME" > env_name
+     fi
+  ) && [ ! -d "$dir" ] && $dryrun mv "${dir}.$tmp" "$dir"
+  # Clean up if we failed or lost the race
+  [ -d "${dir}.$tmp" ] && rm -rf "${dir}.$tmp"
+done
+rm $tmp/tags
+
+# List the installed tags
+(
+  cd "$ROOT/tag"
+  ls -dr d_* > "$STATE/$tmp/d_list" || true
+  ls -dr w_* > "$STATE/$tmp/w_list" || true
+)
+
+# Make the latest links if needed
+for type in d w; do
+  [ -f "$tmp/${type}_list" ] || continue
+  latest=$(head -n 1 "$tmp/${type}_list")
```

Comment:
That's what the `-r` is for on the `ls` command.  It's lexicographically reversed, so this must be the latest.

Reply:
This is true, as long as we don't add anything unexpected or change label layout a bit. This is probably a reasonable assumption, but I just wanted to highlight it, because it could be something we forget about in some future discussion on changing tag names or something.

Reply:
I'd hope that the `d_*` and `w_*` releases would always stay the same; if we needed to change layout, we should use a different prefix.  But I'll add a comment to that effect.